crates/walrus-sdk/src/client/byte_range_read_client.rs

```diff
@@ -0,0 +1,415 @@
+    tracing::debug!("primary_sliver_size: {:?}", primary_sliver_size);
+
+    assert!(byte_length > 0, "byte_length must be greater than 0");
+
+    // Calculate which slivers contain the byte range
+
+    // First sliver that contains start_byte_position
+    let start_sliver_index = start_byte_position / primary_sliver_size;
+
+    // Each systematic (primary) sliver contains primary_sliver_size bytes of the original blob
+    // Note that we subtract 1 from the end byte position because we want to include the last
+    // byte in the last sliver.
+    let end_byte_position = start_byte_position + byte_length - 1;
+
+    // Last sliver that contains end_byte_position
+    let end_sliver_index = end_byte_position / primary_sliver_size;
+
+    // We need to retrieve enough slivers for decoding (n_systematic)
+    // Prioritize slivers starting from start_sliver_index
+    let sliver_indices: Vec<SliverIndex> = (start_sliver_index..=end_sliver_index)
+        .map(|idx| SliverIndex::new(u16::try_from(idx).expect("sliver index should be valid")))
+        .collect();
+
+    tracing::debug!(
+        "byte range {}-{} spans slivers {}-{}, requesting {} slivers",
+        start_byte_position,
+        end_byte_position,
+        start_sliver_index,
+        end_sliver_index,
+        sliver_indices.len()
+    );
+
+    let new_start_byte_position = start_byte_position - start_sliver_index * primary_sliver_size;
+
+    Ok((new_start_byte_position, sliver_indices))
+}
+
+/// Constructs the requested data from the slivers and copy the data directly into the
+/// final result buffer.
+fn construct_requested_data_from_slivers(
+    slivers: &[SliverData<Primary>],
+    new_start_byte_position: usize,
+    byte_length: usize,
+) -> ClientResult<Vec<u8>> {
+    // Copy data directly into the final result buffer, skipping unnecessary bytes
+    let mut result = Vec::with_capacity(byte_length);
+    let mut bytes_to_skip = new_start_byte_position;
+    let mut bytes_remaining = byte_length;
+
+    for sliver in slivers {
+        let data = sliver.symbols.data();
+
+        // Skip entire sliver if we haven't reached the start position yet
+        if bytes_to_skip >= data.len() {
+            return Err(ClientError::from(ClientErrorKind::ByteRangeReadError(
+                format!(
+                    "start byte position should be within the first read sliver. start \
+                        position: {}, sliver size: {}",
+                    new_start_byte_position,
+                    data.len(),
+                ),
+            )));
+        }
+
+        // Copy the needed bytes directly into result
+        let start_in_sliver_inclusive = bytes_to_skip;
+        let end_in_sliver_exclusive = (start_in_sliver_inclusive + bytes_remaining).min(data.len());
+
+        // Copy only the needed bytes directly into result
+        result.extend_from_slice(&data[start_in_sliver_inclusive..end_in_sliver_exclusive]);
```

Comment:
Remark: When implementing some optimizations for the encoding, I noticed lower performance of the `extend` functions compared to something like `copy_from_slice`. That said, this might not be practically relevant here.